### DISCUSSION TITLE
Netdata with Kubernetes - How to mount Disks as described in Issue #5071

### DISCUSSION THREAD

**Discussion by Garahk** [netdata/netdata#13425] [Q&A] [answered]
Hello all,

Based on what is described in issue #5071 to mount the volumes of the host when deploying Netdata in a docker container:

> As far as `diskspace` plugin reads mount list from `/host/proc/self/mountinfo` and get statistics using `statvfs()` you need to mount a volume from your host in order to be able to monitor it. As @autogun already mentioned, you can use `-v /:/host:ro` to do it for `/`.

![image](https://user-images.githubusercontent.com/99432183/180506699-188c0d85-b136-4994-a558-a6a36b654995.png)

What would be the procedure to achieve this in a Kubernetes deployment when deploying Netdata with helm chart?

**Comment by BillKalaitzis** [comment, marked as answer]
Hello @Garahk,

Looking at the linked issue makes me believe that we should add support for `extraVolumes` and `extraVolumeMounts` on our helmchart in order to be able to go by the suggested approach.
I've created a WIP feature branch in here: https://github.com/BillKalaitzis/helmchart/tree/feat/support-extra-volumes, where we add support for the aforementioned directives.

Could you please try to upgrade your helmchart from this repo and branch and see if it works?
You'll also need to append the following under your `child` section in your `values.yaml` (ofc, feel free to alter the actual `hostPath` and `mountPaths` to suit your needs):
```
child:
  extraVolumeMounts:
    - name: root
      mountPath: /host/root
      readOnly: true
  extraVolumes:
    - name: root
      hostPath:
        path: /
```

I will also proceed with testing this more thoroughly on Monday, but would love to hear your feedback as well!

**Reply by Garahk** [reply]
Hello @BillKalaitzis , appreciate your prompt response, am out today, but let me try it tomorrow and I'll get back to you with feedback

**Reply by Garahk** [reply]
Hello @BillKalaitzis , I tested it a couple hours ago a few times with the default configuration you provided, for the `hostPath` and `mountPaths`, however I keep getting the same list under disk, my intention is just to be able to see the volumes under `/` of the host, am I doing something wrong or missing anything?

**Reply by Garahk** [reply]
Captures here:

![image](https://user-images.githubusercontent.com/99432183/180828338-19e97977-6381-49ca-96eb-f480ac0252df.png)

![image](https://user-images.githubusercontent.com/99432183/180828423-bbfd4a92-1214-43cb-a542-6859c9231202.png)

Would it change if this is a Kind type of cluster, I guess it would..

**Reply by ilyam8** [reply]
Added `child.extraHostVolumeMounts` in https://github.com/netdata/helmchart/pull/293.

**Comment by Garahk** [comment]
@BillKalaitzis , I finally understood the way it works, and thought of changing the `mountPath` to `/` so that I would get all volumes (I can discard later):

```
  extraVolumeMounts:
    - name: root
      mountPath: /host/
      readOnly: true
  extraVolumes:
    - name: root
      hostPath:
        path: /
```

I might discard later, but for now, this is what I got:

![image](https://user-images.githubusercontent.com/99432183/180904741-6b06be9b-e293-49ff-9f4b-8edd5036ce27.png)

![image](https://user-images.githubusercontent.com/99432183/180904823-e34c1f18-f331-4852-b4da-bda5e849dd3d.png)

Result is very good.

**Comment by BillKalaitzis** [comment]
Hello @Garahk, glad that you figured it out!

Now that you've mounted the entire `/`,  the disk list will get a bit messy especially thanks to the multiple `shm/tmpfs` mounts.
You can feel free to blacklist specific disks / paths you wouldn't like to monitor via the plugin's `exclude` filters.Have a look in here for more info: https://learn.netdata.cloud/docs/agent/collectors/diskspace.plugin.

In the meantime, I'll proceed with creating a PR to merge the related helmchart changes on our upstream repo.

**Reply by ilyam8** [reply]
@BillKalaitzis I went forward and created a PR - https://github.com/netdata/helmchart/pull/293. It is only one option instead of two (extraHostVolumeMounts vs extraVolumes + extraVolumeMounts).

**Reply by Garahk** [reply]
@BillKalaitzis , Recently a member of my team requested to monitor the `opt` and `boot` partitions, located in `/`:

![image](https://user-images.githubusercontent.com/99432183/187514459-4b21efb1-414a-41a4-a27d-62d2bc2e225b.png)

My current configuration in `ExtraVolumes` and `ExtraVolumeMounts` is the following:

```
  extraVolumeMounts:
    - name: root
      mountPath: /host/
      readOnly: true

  extraVolumes:
    - name: root
      hostPath:
        path: /
```

Which provides me with the following disk mounts:

![image](https://user-images.githubusercontent.com/99432183/187514931-5859d14d-1437-4a09-a48b-50597618dadc.png)

And am only excluding the following paths:

```
        [plugin:proc:diskspace]
            exclude space metrics on paths = /proc/* /sys/* /var/run/user/* /run/user/* /snap/* /var/lib/docker/* /host/run/* /host/var/lib/*

```

As you may notice, am not excluding `opt` or `boot`.

What can I change or do so that I can add not just `opt` or `boot`, but to add any additional volume from the host?

Thank you in advance.

**Reply by BillKalaitzis** [reply]
Hey @Garahk,

Are the `/opt` and `/boot` mounted as separate partitions on your system?
Can you please give me the output of the command `mount` to have a look at your mountpoints?

**Reply by Garahk** [reply]
@BillKalaitzis , adding the `mount` output here:
[mount.log](https://github.com/netdata/netdata/files/9456787/mount.log)

**Reply by BillKalaitzis** [reply]
Hey @Garahk,

Thank you for the logs.
Can you please also send me the output of the following command (should be run from inside the netdata container): `curl -s http://localhost:19999/netdata.conf`

Note: Please make sure to filter out any sensitive information from the netdata.conf output that you'll be uploading.

**Reply by Garahk** [reply]
Hello @BillKalaitzis , added here the netdata.conf
[netdata.zip](https://github.com/netdata/netdata/files/9456979/netdata.zip)

**Reply by BillKalaitzis** [reply]
I've had a look at the `mount` logs and it seems that the `/opt` directive isn't there, which means that it's not a separate partition, but instead it's merged with the root partition. Can you please verify that? 

On the other hand, `/boot` is indeed present on the `mount` log. The fact that there's not a section about `[plugin:proc:diskspace:/host/boot]` into your `netdata.conf` makes me believe that for whatever reason Netdata doesn't recognise it. Can you please enter the Netdata container, issue the following command and post the logs?
```
$ cat /proc/1/mountinfo
```

Also, you could try to manually append the diskspace entry for `/host/boot`, to see whether it will yield any results?
To do this, you'll need to add the following section on your netdata.conf:
```
[plugin:proc:diskspace:/host/boot]
        space usage = yes
        inodes usage = yes
```

**Reply by Garahk** [reply]
hello @BillKalaitzis , apologies for the delay in answering, adding here the contents of `$ cat /proc/1/mountinfo`:
[mountinfo.log](https://github.com/netdata/netdata/files/9470385/mountinfo.log)

And I also tested what you mentioned, by adding the 2 following sections:

```
[plugin:proc:diskspace:/host/boot]
        space usage = yes
        inodes usage = yes
```
And:

```
  extraVolumeMounts:
    - name: root
      mountPath: /host/
      readOnly: true
    - name: boot
      mountPath: /host/boot
      readOnly: true

  extraVolumes:
    - name: root
      hostPath:
        path: /
    - name: boot
      hostPath:
        path: /boot
```

Adding both sections, `extraVolumes/extraVolumeMounts` and in `netdata.conf` provides the `boot` storage metrics, did I understand correctly that both sections were needed? or just the plugin in `netdata.conf` was supposed to make it work?

Note that if I don't explicitly add the `boot` section as a mount it won't be added in the disks section.

![image](https://user-images.githubusercontent.com/99432183/187932195-cf45741f-eb22-4928-8718-679a99688590.png)

Please let me know your thoughts.

**Reply by BillKalaitzis** [reply]
Hey @Garahk, sorry for the late reply.
In my original message I was suggesting to only set the `[plugin:proc:diskspace:/host/boot]` section, just to see if it would help.
If it wouldn't, then the follow-up suggestion would be to try and mount the `/boot` partition on the Netdata pod.
You've figured this already by yourself, so now we're good 😄

**Reply by Garahk** [reply]
Hello @BillKalaitzis , No worries, thought you might have been busy with other matters, and thank you for confirmation that the suggestion was good. I really appreciate the great support your team always provide.
